util/bloom_test.cc

```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+
+  // Number 500,000 is carefully chosen so that memory reserved before banding
+  // won't cause a cache full under strict capacity limit but will after.
+  // Therefore it will trigger a fallback due to banding portion being too large
+  // during FilterBitsBuilder::Finish().
+  //
+  // Calculation:
+  // Before banding: hash_entries_.size() * hash_entries_[0] ≈ 500,000 * 8 =
+  // 4,000,000 > 3MB Banding'size ≈ 10,437,168 > 9MB After banding: 3MB + 9MB >
+  // 8MB
+  //
+  // Number 50,000 is a small number so there will not be any cache full hence
+  // fallback.
+  //
+  // Calculation:
+  // Before banding: hash_entries_.size() * hash_entries_[0] ≈ 50,000 * 8 =
+  // 400,000 < 0.4MB Banding'size ≈ 1,034,288 < 2MB After banding: 0.4 MB + 2MB
+  // < 8MB
+  for (std::size_t num_entries : {500000, 50000}) {
+    bool will_fall_back = (num_entries == 500000);
+
+    BlockBasedTableOptions table_options;
+    LRUCacheOptions lo;
+    lo.capacity = kCacheCapacity;
+    lo.num_shard_bits = 0;  // 2^0 shard
+    lo.strict_capacity_limit = true;
+    std::shared_ptr<Cache> cache(NewLRUCache(lo));
+    table_options.block_cache = cache;
+    table_options.filter_policy.reset(
+        new BloomFilterPolicy(FLAGS_bits_per_key, mode));
+    table_options.reserve_table_builder_memory = true;
+
+    FilterBuildingContext ctx(table_options);
+
+    ASSERT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+    std::unique_ptr<FilterBitsBuilder> filter_bits_builder(
+        table_options.filter_policy->GetBuilderWithContext(ctx));
+
+    char key_buffer[sizeof(int)];
+    for (std::size_t i = 0; i < num_entries; ++i) {
+      filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
+    }
+    // 0xffffffffU is max size supported by implementation for buf length
+    std::unique_ptr<const char[]> buf(new char[0xffffffffU]);
+    Slice filter = filter_bits_builder->Finish(&buf);
+    // To verify Ribbon Filter fallbacks to Bloom Filter properly
+    // based on cache reservation result
+    // See BloomFilterPolicy::GetBloomBitsReader re: metadata
+    // -1 = Marker for newer Bloom implementations
+    // -2 = Marker for Standard128 Ribbon
+    if (will_fall_back) {
+      EXPECT_EQ(filter.data()[filter.size() - 5], static_cast<char>(-1));
+    } else {
+      EXPECT_EQ(filter.data()[filter.size() - 5], static_cast<char>(-2));
+    }
+
+    // Memory reserved now:  mutable_buffer_'s size (approximated by
+    // filter.size())
+    memory_reserved = filter.size();
+    num_dummy_entries = static_cast<std::size_t>(
+        std::ceil(memory_reserved / static_cast<double>(kSizeDummyEntry)));
+    EXPECT_GE(cache->GetPinnedUsage(), num_dummy_entries * kSizeDummyEntry);
+    EXPECT_LT(cache->GetPinnedUsage(),
+              (num_dummy_entries + 1) * kSizeDummyEntry);
+
+    filter_bits_builder.reset();
+    EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+  }
+}
+
+class CustomFilterPolicy : public FilterPolicy {
```

Comment:
Not sure why this class is copy-pasted. IMHO this PR doesn't do anything to make it susceptible to failure under a wrapper custom filter policy but not under a built-in filter policy, so this custom filter policy seems like extra, unnecessary test code (that we have to maintain forever).